crypto/fipsmodule/rand/entropy/tree_drbg_jitter_entropy.c

```diff
@@ -0,0 +1,465 @@
+  if (tree_jitter_drbg == NULL) {
+    abort();
+  }
+
+  // Initialize the global seed DRBG if haven't already.
+  CRYPTO_once(global_seed_drbg_once_bss_get(), tree_jitter_initialize_once);
+
+  // Initialize the per-thread seed DRBG.
+  uint8_t seed_drbg[CTR_DRBG_ENTROPY_LEN];
+  CRYPTO_STATIC_MUTEX_lock_write(global_seed_drbg_lock_bss_get());
+  tree_jitter_drbg_derive_seed(*global_seed_drbg_bss_get(), seed_drbg);
+  CRYPTO_STATIC_MUTEX_unlock_write(global_seed_drbg_lock_bss_get());
+
+  if (!CTR_DRBG_init(&(tree_jitter_drbg->drbg), seed_drbg, NULL, 0)) {
+    abort();
+  }
+  OPENSSL_cleanse(seed_drbg, CTR_DRBG_ENTROPY_LEN);
+
+  tree_jitter_drbg->generate_calls_since_seed = 0;
+  tree_jitter_drbg->max_generate_calls = TREE_JITTER_THREAD_DRBG_MAX_GENERATE;
+  tree_jitter_drbg->reseed_calls_since_initialization = 0;
+  uint64_t current_generation_number = 0;
+  if (CRYPTO_get_ube_generation_number(&current_generation_number) != 1) {
+    tree_jitter_drbg->ube_protection = 0;
+    tree_jitter_drbg->generation_number = 0;
+  } else {
+    tree_jitter_drbg->ube_protection = 1;
+    tree_jitter_drbg->generation_number = current_generation_number;
+  }
+
+  entropy_source->state = tree_jitter_drbg;
+
+  return 1;
+}
+
+#if defined(_MSC_VER)
+#pragma section(".CRT$XCU", read)
+static void tree_jitter_free_global_drbg(void);
+static void windows_install_tree_jitter_free_global_drbg(void) {
+  atexit(&tree_jitter_free_global_drbg);
+}
+__declspec(allocate(".CRT$XCU")) void(*tree_jitter_drbg_destructor)(void) =
+    windows_install_tree_jitter_free_global_drbg;
+#else
+static void tree_jitter_free_global_drbg(void) __attribute__ ((destructor));
+#endif
+
+// The memory life-time for thread-local seed DRBGs is handled differently
+// compared to the global seed DRBG (and Jitter Entropy instance). The frontend
+// DRBG thread-local destuctors will invoke |tree_jitter_free_thread_drbg| using
+// their reference to it. The global seed DRBG and Jitter Entropy instance will
+// be released by a destructor. This ensures that the global seed DRBG life-time
+// extends to the entire process life-time if the lazy initialization happened.
+// Obviously, any dlclose on AWS-LC will release the memory early but that's
+// correct behaviour.
+
+// tree_jitter_free_global_drbg frees the memory allocated for the global seed
+// DRBG and Jitter Entropy instance.
+static void tree_jitter_free_global_drbg(void) {
+
+  CRYPTO_STATIC_MUTEX_lock_write(global_seed_drbg_lock_bss_get());
+
+  struct tree_jitter_drbg_t *global_tree_jitter_drbg = *global_seed_drbg_bss_get();
+  if (*global_seed_drbg_bss_get() == NULL) {
+    return;
+  }
+
+  if (global_tree_jitter_drbg->is_global != 1) {
+    // Should not happen.
+    abort();
+  }
+
+  jent_entropy_collector_free(global_tree_jitter_drbg->jitter_ec);
+  OPENSSL_free(global_tree_jitter_drbg);
+
+  *global_seed_drbg_bss_get() = NULL;
+
+  CRYPTO_STATIC_MUTEX_unlock_write(global_seed_drbg_lock_bss_get());
```

Comment:
How does CRYPTO_once work? Will it see global_seed_drbg_bss_get is null and run again? Also if that does happen after this do we need to free this global state again? Should this just not unlock here?

Reply:
> Will it see global_seed_drbg_bss_get is null and run again

No it won't. Only if you reset the lock, which you shouldn't do.

I check whether the global is NULL before using it throughout. At least supposed to :)

But this is just hard to ensure if consumers doesn't exit applications gracefully. When dealing with concurrency you at least need to make some assumptions and then just abort if those are not met. IMO.

Reply:
> No it won't. Only if you reset the lock, which you shouldn't do.

Isn't that what this line is doing?

Reply:
The lock used is `global_seed_drbg_once` and the global is initialized per:
```
CRYPTO_once(global_seed_drbg_once_bss_get(), tree_jitter_initialize_once);
```
invoked from the interface function `tree_jitter_initialize()`, that is called for every new thread. No matter how many times this function is called, `tree_jitter_initialize_once()` will only ever run once.